mitosheet/mitosheet/tests/step_performers/test_user_defined_edit.py

```diff
@@ -0,0 +1,189 @@
+    df['B'] = tmp
+    df.columns = ['B', 'A']
+    return df
+
+
+USER_DEFINED_IMPORT_TESTS = [
+    (
+        [
+            START_DF
+        ],
+        "no_change",
+        {
+            'random_name': 'df1'
+        },
+        [
+            START_DF
+        ]
+    ),
+    (
+        [
+            START_DF
+        ],
+        "add_one",
+        {
+            'df': 'df1'
+        },
+        [
+            START_DF + 1
+        ]
+    ),
+    (
+        [
+            START_DF
+        ],
+        "add_param",
+        {
+            'df': 'df1',
+            'add_val': '2'
+        },
+        [
+            START_DF + 2
+        ]
+    ),
+    (
+        [
+            START_DF
+        ],
+        "add_col",
+        {
+            'df': 'df1',
+        },
+        [
+            ADDED_COL_DF
+        ]
+    ),
+    (
+        [
+            START_DF
+        ],
+        "del_col",
+        {
+            'df': 'df1',
+        },
+        [
+            START_DF.drop('A', axis=1)
+        ]
+    ),
+    (
+        [
+            START_DF
+        ],
+        "rename_all_columns",
+        {
+            'df': 'df1',
+        },
+        [
+            RENAMED_COL_DF
+        ]
+    ),
+    (
+        [
+            START_DF
+        ],
+        "reorder_columns",
+        {
+            'df': 'df1',
+        },
+        [
+            REORDER_COL_DF
+        ]
+    ),
+]
+@pytest.mark.parametrize("input_dfs, edit_name, edit_params, output_dfs", USER_DEFINED_IMPORT_TESTS)
+def test_userdefinedimport(input_dfs, edit_name, edit_params, output_dfs):
+    mito = create_mito_wrapper(*[df.copy(deep=True) for df in input_dfs], editors=[no_change, add_one, add_param, add_col, del_col, rename_all_columns, reorder_columns])
+    mito.user_defined_edit(edit_name, edit_params)
+
+    assert len(mito.dfs) == len(output_dfs)
+    for actual, expected in zip(mito.dfs, output_dfs):
+        assert actual.equals(expected)
+
+# TODO: make sure we can do another edit after it
+
+def no_params() -> pd.DataFrame:
+    import pandas as pd
+    df = pd.DataFrame({'A': [1]})
+    return df
+
+def wrong_first_param(df: int) -> pd.DataFrame:
+    import pandas as pd
+    return pd.DataFrame({'A': [1, 2, 3]})
+
+from mitosheet.extensions.v1 import ColumnHeader
+def unqualified_column_header(ch: ColumnHeader) -> pd.DataFrame:
+    import pandas as pd
+    return pd.DataFrame({'A': [1, 2, 3]})
+
+def test_user_defined_edit_errors_with_no_params():
+    with pytest.raises(ValueError) as e_info:
+        create_mito_wrapper(editors=[no_params])
+    
+def test_user_defined_edit_errors_with_wrong_first_param_type():
+    with pytest.raises(ValueError) as e_info:
+        create_mito_wrapper(editors=[wrong_first_param])
+
+def test_user_defined_edit_with_unqualified_column_header_errors():
+    with pytest.raises(ValueError) as e_info:
+        create_mito_wrapper(editors=[unqualified_column_header])
+
+def test_user_defined_edit_optimizes():
```

Comment:
Can we add a test to make sure if you add a column to a dataframe, perform an edit to that dataframe and then delete the column, the code does not optimize out the add column. etc